blog/_posts/2019-09-12-tips-on-moving-your-dev-env-from-docker-compose-to-kubernetes.md

```diff
@@ -0,0 +1,173 @@
+
+1. I have a Docker-Compose file with several apps
+
+2. Kubernetes represents a single running app as a Pod
+
+3. Therefore, I should convert each app config to a Pod config and see if it works
+
+Don't do this!
+
+A pod is a good building block but a terrible dev experience.
+
+When you replace a Pod, Kubernetes will, in serial, signal the process to shut
+down, gracefully wait for it to exit, free all the pod resources, and then start
+your new Pod. This is too slow.
+
+But with a bit of planning, Kubernetes can parallelize these steps. It can run
+the new process and the old process side-by-side, directing traffic to the new
+server while the old one is shutting down.
+
+## Key Insight: Play to Kubernetes' Strengths
+
+Kubernetes is good at incremental updates, components, and reusability. How can
+we take advantage of that?
+
+### Step 1) Start with a docker-compose file
+
+You can look at the one in our [sample repo](https://github.com/windmilleng/express-redis-docker)
+
+### Step 2) Run `kompose convert`
+
+[Kompose](https://github.com/kubernetes/kompose) is a tool that helps convert Docker-Compose configs to Kubernetes configs.
+
+Kompose will create a Deployment YAML and Service YAML for each app.
+
+They won't work. Don’t Panic! That’s OK.
+
+Kompose has lots of bugs. It's still using beta Kubernetes Deployments by default
+and seems unmaintained. For most projects, it won't be a complete solution. But
+it's still a great start to build from.
+
+### Step 3) Get a “base” service working
+
+Find a base service that doesn't depend on any other services.
+
+Deploy it to a local Kubernetes cluster. Tweak the Kubernetes configs and keep
+`kubectl apply`-ing until it's running. Use
+[`kubectl port-forward`](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/)
+to connect the service to a port on your machine, and test it manually with a
+browser.
+
+If you get stuck, [`kubespy trace` ](https://www.pulumi.com/blog/kubespy-trace-a-real-time-view-into-the-heart-of-a-kubernetes-service/)
+is a great tool for checking common misconfiguration.
+
+### Step 4) Get a service the depends on “base” working
+
+Find a service that only depends on your base service.
+
+Keep `kubectl apply`-ing it until it's running OK, just like you did in Step 3.
+
+Repeat until all services work!
+
+### Step 5) Throw docker-compose away
+
+This is a good time to throw a party.
+
+## How Tilt Can Help
+
+One of the big reasons we made Tilt is to make it easier to iteratively hack on
+YAML files. The workflow I use:
+
+- Create a Tiltfile with the new K8s configs behind an `if` branch, so that I
+  can check them in one at a time without affecting teammates using Docker-Compose.
+- Leave Tilt running while I hack on YAML files, so that Tilt continuously redeploys them
+- Add port-forwards to the Tiltfile, so that Tilt will automatically them on each redeploy.
```

Comment:
```suggestion
- Add port-forwards to the Tiltfile, so that I can test my services manually in the browser without wrestling with `kubectl port-forward`
```

Reply:
done